### PR TITLE
Migrate @storybook/components to TS

### PR DESCRIPTION
**Pull request by gaetanmaisse**
Issue: https://github.com/storybooks/storybook/issues/5030

## What I did

 - [x] 🏗 Migrate `@storybook/components` to TypeScript
 - [x] 🔧 Add some types and interfaces and use them to type codebase 
 - [x] ⚔️ Fight with TS compiler about anonymous functions, see https://github.com/storybooks/storybook/pull/6095#issuecomment-477480930 and https://github.com/storybooks/storybook/pull/6095/commits/5fa1be5447c657705879604c4eda7fe6fc13c0d5
 - [x] ♻️ Localy disable a md linter rule (as we need to do what is done) to (try to) fix `Lint (Storybook)` job

## TODO in another PR

 - 🛠 Refactor some imports/exports
 - 📦 Use these typings in addons

### CODE REVIEW

**Comment by ndelangen** [comment]
Mind this PR:
https://github.com/storybooks/storybook/pull/5402

You may want to base on that, it's close to being merged.

File: lib/components/package.json
Comment (gaetanmaisse) [inline lib/components/package.json]: Maybe it should be placed in `dependencies `

**Comment by ndelangen** [comment]
@gaetanmaisse how's this going? Do you need assistance?

This would be an immense step forward for our typescript migration epic.

**Comment by gaetanmaisse** [comment]
@ndelangen I migrate codebase to TS and the only thing still in progress before I can put this PR in review is fixing snapshot testing: as TS compiler generates js sources that are not the same as Babel ones it makes tests failed  (some functions are anonymous in TS build). 

For instance:

Babel generated source for Scrollbar:
```js
var ScrollArea = function ScrollArea(_ref4) {
  var children = _ref4.children,
      vertical = _ref4.vertical,
      horizontal = _ref4.horizontal,
      props = _objectWithoutProperties(_ref4, ["children", "vertical", "horizontal"]);

  return _react.default.createElement(_react.Fragment, null, _ref5, _react.default.createElement(Scroll, _extends({
    vertical: vertical,
    horizontal: horizontal
  }, props), children));
};

exports.ScrollArea = ScrollArea;
```

TS generated source for Scrollbar:
```js
exports.ScrollArea = function (_a) {
    var children = _a.children, vertical = _a.vertical, horizontal = _a.horizontal, props = __rest(_a, ["children", "vertical", "horizontal"]);
    return (react_1.default.createElement(react_1.Fragment, null,
        react_1.default.createElement(theming_1.Global, { styles: ScrollAreaStyles_1.getScrollAreaStyles }),
        react_1.default.createElement(Scroll, __assign({ vertical: vertical, horizontal: horizontal }, props), children)));
};
```

With Babel compiler `ScrollArea` tag name will be used in snapshots but with TS it will be a default `Component` tag.

Set `"module":"es6"` in tsconfig fixes this problem but causes build error as externals libs can not be imported anymore... 🤷‍♂️ still investigating on this

File: lib/components/src/brand/StorybookIcon.tsx
Comment (gaetanmaisse) [inline lib/components/src/brand/StorybookIcon.tsx]: `title` is not a valid attribute of `svg` tag but an inner tag

File: lib/components/src/ActionBar/ActionBar.tsx
Comment (benoitdion) [inline lib/components/src/ActionBar/ActionBar.tsx]: nit: my personal preference is to use the `FC` short-hand since that makes the lines shorter

File: lib/components/src/ScrollArea/ScrollAreaStyles.tsx
Comment (kroeder) [inline lib/components/src/ScrollArea/ScrollAreaStyles.tsx]: Why where you in need of casting `any` here?

File: lib/components/src/form/input/input.tsx
Comment (kroeder) [inline lib/components/src/form/input/input.tsx]: Is there anything we can do about all these `any` for `styled`?

File: lib/components/src/tabs/tabs.tsx
Comment (kroeder) [inline lib/components/src/tabs/tabs.tsx]: Isn't this just `e:Event` if you don't want to be more specific?

File: lib/components/src/tooltip/Tooltip.tsx
Comment (kroeder) [inline lib/components/src/tooltip/Tooltip.tsx]: Can you explain what's going on here? Never saw an interface using a `keyof Color` type

File: lib/components/src/tooltip/WithTooltip.tsx
Comment (kroeder) [inline lib/components/src/tooltip/WithTooltip.tsx]: Why do you needed to cast `this as any` ?

**Review by kroeder — APPROVED** [review]
Good job! 🚀

File: lib/components/src/ActionBar/ActionBar.stories.tsx
Comment (leoyli) [inline lib/components/src/ActionBar/ActionBar.stories.tsx]: `JSX` comes from the global, which is subject to change and may not be future proof...  Here is [the reason with some tips](https://medium.com/@martin_hotell/10-typescript-pro-tips-patterns-with-or-without-react-5799488d6680#58e3) that I found super useful.

File: lib/components/src/ActionBar/ActionBar.tsx
Comment (leoyli) [reply lib/components/src/ActionBar/ActionBar.tsx]: But with IDE this should not be too problematic I believe. I personally think shorthand is evil since it is not really reader friendly.

File: lib/components/src/tooltip/WithTooltip.tsx
Comment (leoyli) [inline lib/components/src/tooltip/WithTooltip.tsx]: maybe use `unknown` instead of `any`?!  Here is a [great explanation around `any`, `never`, and `unknown`](https://blog.logrocket.com/when-to-use-never-and-unknown-in-typescript-5e4d6c5799ad) and when to use them. In general, I only found typing function with arbitrary numbers of arguments will use `any`.

File: lib/components/src/tooltip/Tooltip.tsx
Comment (leoyli) [reply lib/components/src/tooltip/Tooltip.tsx]: `keyof` is a method(? or lookup type?) to get all member names in a type.  This gives a string subset typed with the names of all property in the `Color` type.  I.e., `keyof Color === 'primary' | 'secondary'| ...`.

**Comment by leoyli** [comment]
Don't have time to dig in too detail but I think it is a great work for sure!  I've attached couple comments and referenced some great articles. Thank you very much!

File: lib/components/src/ActionBar/ActionBar.tsx
Comment (leoyli) [reply lib/components/src/ActionBar/ActionBar.tsx]: This comment reminds me that typing in React is really not an easy job... Maybe not so interesting, but I think some component could be more restrict to its children with this generic type recipe: `type FCNoChildren<P> = FunctionComponent<{ children?: never } & P>;` for those component should not ever have a children (i.e. always inlined).

Comment (gaetanmaisse) [reply lib/components/src/ActionBar/ActionBar.tsx]: I agree with @leoyli on:
> shorthand is evil since it is not really reader friendly

Especially for react newbie like me.

Comment (benoitdion) [reply lib/components/src/ActionBar/ActionBar.tsx]: I almost always prefer longer names but this is one of the few exceptions since `FC`s are all over the place. It helps avoid line wraps when using descriptive component and prop names. I'd use it with `React.FC` though, importing just `FC` does feel like a little too much.

With all that said, if the convention so far has been to import `FunctionComponent` it makes sense to continue doing that.

File: lib/components/src/tabs/tabs.tsx
Comment (gaetanmaisse) [reply lib/components/src/tabs/tabs.tsx]: Hmm yep I missed that but know that I look at this line I don't really understand it, what is
`e => e.preventDefault() || actions.onSelect(id)` supposed to do? because `e => e.preventDefault()` is always defined 🤔 

 cc @ndelangen 😇

File: lib/components/src/ActionBar/ActionBar.stories.tsx
Comment (gaetanmaisse) [reply lib/components/src/ActionBar/ActionBar.stories.tsx]: I didn't know about that, good article!

File: lib/components/src/tooltip/WithTooltip.tsx
Comment (gaetanmaisse) [reply lib/components/src/tooltip/WithTooltip.tsx]: Fixed! Just as for `placement` I have used types from `popper.js` (the underlying library used for tooltips) so no more `any` here. I sometimes use `unknown` but it's not a reflex yet, soon ;)

File: lib/components/tsconfig.json
Comment (gaetanmaisse) [inline lib/components/tsconfig.json]: @kroeder  I had to use `types` to includes types of some library as it's done like this in "parent" [`tsconfig.json`](https://github.com/storybooks/storybook/blob/next/tsconfig.json), is there a reason to do it instead of letting TS compiler use all types available in `/node_modules/@types/` based on _@types, typeRoots and types_ of https://www.typescriptlang.org/docs/handbook/tsconfig-json.html

File: lib/components/src/form/input/input.tsx
Comment (gaetanmaisse) [reply lib/components/src/form/input/input.tsx]: I tried to but have no solution for now... 😞

File: lib/components/src/tooltip/WithTooltip.tsx
Comment (gaetanmaisse) [reply lib/components/src/tooltip/WithTooltip.tsx]:
```
Error:(139, 16) TS2339: Property 'unbind' does not exist on type '_ReactLifeCycleFunctionsThisArguments<WithTooltipPureProps, {}>'.
```
I think `WithTooltip` component need to be refactored in order to make these `bind` and `unbind` functions class properties or something like that. I keep it in mind for another PR.

File: lib/components/package.json
Comment (ndelangen) [inline lib/components/package.json]: must be updated to 24.7.1

File: lib/components/src/ActionBar/ActionBar.tsx
Comment (ndelangen) [reply lib/components/src/ActionBar/ActionBar.tsx]: Let's keep the import as `FunctionComponent`. If desired, we can create the shorthand locally in the module.

This will allow newcomers to better understand what's going on, and find the proper documentation more easily.

File: lib/components/src/bar/separator.tsx
Comment (ndelangen) [inline lib/components/src/bar/separator.tsx]: any[] would be ReactElement[]

File: lib/components/src/placeholder/placeholder.tsx
Comment (ndelangen) [inline lib/components/src/placeholder/placeholder.tsx]: Is this still valid after the change to use babel for compilation?

File: lib/components/src/syntaxhighlighter/syntaxhighlighter.tsx
Comment (ndelangen) [inline lib/components/src/syntaxhighlighter/syntaxhighlighter.tsx]: add as static property?

File: lib/components/src/tabs/tabs.tsx
Comment (ndelangen) [reply lib/components/src/tabs/tabs.tsx]: `e.preventDefault()` will return `undefined` always.

so `e.preventDefault() || actions.onSelect(id)` will call both.

it's the same as

```
(e: any) => {
  e.preventDefault();
  actions.onSelect(id);
}
```

File: lib/components/src/tabs/tabs.tsx
Comment (ndelangen) [inline lib/components/src/tabs/tabs.tsx]: can be set as a static property?

File: lib/components/src/tooltip/WithTooltip.tsx
Comment (ndelangen) [reply lib/components/src/tooltip/WithTooltip.tsx]: 👍

File: lib/components/src/tabs/tabs.tsx
Comment (gaetanmaisse) [reply lib/components/src/tabs/tabs.tsx]: Hmm I did not find an easy way to do it as `Tabs` is created using `React.memo`

File: lib/components/src/placeholder/placeholder.tsx
Comment (gaetanmaisse) [reply lib/components/src/placeholder/placeholder.tsx]: I will revert my commit to try ;)

**Comment by gaetanmaisse** [comment]
@ndelangen  @shilman I think we will be good to go with this one after a final check on CI results (still running when I'm writing this). Will this be part of 5.1.X or should we wait before merging?

Then I will continue with "Part II" :
 - 🛠 Refactor some imports/exports
 - 📦 Use these typings in addons
 - 🐎 Check if build can be faster

File: lib/components/src/placeholder/placeholder.tsx
Comment (gaetanmaisse) [reply lib/components/src/placeholder/placeholder.tsx]: 🎉

File: lib/components/src/ActionBar/ActionBar.stories.tsx
Comment (mohsinulhaq) [inline lib/components/src/ActionBar/ActionBar.stories.tsx]: this space removal doesn't seem intentional. or is it?

Comment (gaetanmaisse) [reply lib/components/src/ActionBar/ActionBar.stories.tsx]: It wasn't, will be back in https://github.com/storybooks/storybook/pull/6500 😉

**Pull request merged**
